### ISSUE TITLE
Add pre-commit

### ISSUE THREAD

**Issue by ZryletTC** [pcdshub/pcds-envs#120]
See if pre-commit can be added to our environment so that the pre-commit hooks can be properly used on the SLAC filesystem.

**Comment by ZryletTC** [comment]
Will have to look more into it but my preliminary test didn't work:
```
(pre-commit) psbuild-rhel7-01:jet_tracking pennebak$ pre-commit run --all-files
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
return code: 128
expected return code: 0
stdout: (none)
stderr:
    error: RPC failed; result=22, HTTP code = 404
    fatal: The remote end hung up unexpectedly
    
Check the log at /reg/neh/home/pennebak/.cache/pre-commit/pre-commit.log
```
<details><summary>Log</summary>
<p>
### version information

```
pre-commit version: 2.2.0
sys.version:
    3.7.6 | packaged by conda-forge | (default, Mar 23 2020, 23:03:20) 
    [GCC 7.3.0]
sys.executable: /reg/neh/home/pennebak/.conda/envs/pre-commit/bin/python
os.name: posix
sys.platform: linux
```

### error information

```
An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
return code: 128
expected return code: 0
stdout: (none)
stderr:
    error: RPC failed; result=22, HTTP code = 404
    fatal: The remote end hung up unexpectedly
    
```

```
Traceback (most recent call last):
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 178, in clone_strategy
    self._shallow_clone(ref, _git_cmd)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 160, in _shallow_clone
    git_cmd('-c', git_config, 'fetch', 'origin', ref, '--depth=1')
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 175, in _git_cmd
    cmd_output_b('git', *args, cwd=directory, env=env)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/util.py", line 156, in cmd_output_b
    raise CalledProcessError(returncode, cmd, retcode, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('/usr/bin/git', '-c', 'protocol.version=2', 'fetch', 'origin', '3.7.9', '--depth=1')
return code: 128
expected return code: 0
stdout: (none)
stderr:
    error: RPC failed; result=22, HTTP code = 404
    fatal: The remote end hung up unexpectedly
    

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/error_handler.py", line 56, in error_handler
    yield
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/main.py", line 392, in main
    return run(args.config, store, args)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/commands/run.py", line 344, in run
    for hook in all_hooks(config, store)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/repository.py", line 206, in all_hooks
    for repo in root_config['repos']
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/repository.py", line 207, in <genexpr>
    for hook in _repository_hooks(repo, store, root_config)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/repository.py", line 182, in _repository_hooks
    return _cloned_repository_hooks(repo_config, store, root_config)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/repository.py", line 148, in _cloned_repository_hooks
    manifest_path = os.path.join(store.clone(repo, rev), C.MANIFEST_FILE)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 182, in clone
    return self._new_repo(repo, ref, deps, clone_strategy)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 139, in _new_repo
    make_strategy(directory)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 180, in clone_strategy
    self._complete_clone(ref, _git_cmd)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 152, in _complete_clone
    git_cmd('fetch', 'origin', '--tags')
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/store.py", line 175, in _git_cmd
    cmd_output_b('git', *args, cwd=directory, env=env)
  File "/reg/neh/home/pennebak/.conda/envs/pre-commit/lib/python3.7/site-packages/pre_commit/util.py", line 156, in cmd_output_b
    raise CalledProcessError(returncode, cmd, retcode, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
return code: 128
expected return code: 0
stdout: (none)
stderr:
    error: RPC failed; result=22, HTTP code = 404
    fatal: The remote end hung up unexpectedly
    

```
</details>

<details><summary>Conda Info</summary>
<p>
```
     active environment : pre-commit
    active env location : /reg/neh/home/pennebak/.conda/envs/pre-commit
            shell level : 3
       user config file : /reg/neh/home/pennebak/.condarc
 populated config files : /reg/g/pcds/pyps/conda/py36/.condarc
                          /reg/neh/home/pennebak/.condarc
          conda version : 4.8.2
    conda-build version : not installed
         python version : 3.7.0.final.0
       virtual packages : __glibc=2.17
       base environment : /reg/g/pcds/pyps/conda/py36  (read only)
           channel URLs : https://conda.anaconda.org/pcds-tag/linux-64
                          https://conda.anaconda.org/pcds-tag/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /reg/g/pcds/pyps/conda/py36/pkgs
                          /reg/neh/home/pennebak/.conda/pkgs
       envs directories : /reg/neh/home/pennebak/.conda/envs
                          /reg/g/pcds/pyps/conda/py36/envs
               platform : linux-64
             user-agent : conda/4.8.2 requests/2.22.0 CPython/3.7.0 Linux/3.10.0-1062.el7.x86_64 rhel/7.3 glibc/2.17
                UID:GID : 16836:1096
             netrc file : None
           offline mode : False

# conda environments:
#
base                     /reg/g/pcds/pyps/conda/py36
dev                      /reg/g/pcds/pyps/conda/py36/envs/dev
pcds-0.1.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.1.0
pcds-0.2.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.2.0
pcds-0.3.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.3.0
pcds-0.4.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.4.0
pcds-0.5.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.5.0
pcds-0.5.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.5.1
pcds-0.5.2               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.5.2
pcds-0.6.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.6.0
pcds-0.6.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-0.6.1
pcds-1.0.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.0.0
pcds-1.1.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.1.0
pcds-1.1.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.1.1
pcds-1.2.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.0
pcds-1.2.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.1
pcds-1.2.2               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.2
pcds-1.2.3               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.3
pcds-1.2.4               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.4
pcds-1.2.5               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.5
pcds-1.2.6               /reg/g/pcds/pyps/conda/py36/envs/pcds-1.2.6
pcds-2.0.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.0.0
pcds-2.0.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.0.1
pcds-2.0.2               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.0.2
pcds-2.0.3               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.0.3
pcds-2.0.4               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.0.4
pcds-2.1.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.0
pcds-2.1.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.1.1
pcds-2.2.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.2.0
pcds-2.3.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.3.0
pcds-2.3.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.3.1
pcds-2.3.2               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.3.2
pcds-2.3.3               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.3.3
pcds-2.3.4               /reg/g/pcds/pyps/conda/py36/envs/pcds-2.3.4
pcds-3.0.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0
pcds-3.0.1               /reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.1
pcds-3.1.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-3.1.0
pcds-3.2.0               /reg/g/pcds/pyps/conda/py36/envs/pcds-3.2.0
pre-commit            *  /reg/neh/home/pennebak/.conda/envs/pre-commit

sys.version: 3.7.0 (default, Jun 28 2018, 13:15:42) 
...
sys.prefix: /reg/g/pcds/pyps/conda/py36
sys.executable: /reg/g/pcds/pyps/conda/py36/bin/python
conda location: /reg/g/pcds/pyps/conda/py36/lib/python3.7/site-packages/conda
conda-build: None
conda-env: /reg/g/pcds/pyps/conda/py36/bin/conda-env
user site dirs: 

CIO_TEST: <not set>
CONDA_DEFAULT_ENV: pre-commit
CONDA_EXE: /reg/g/pcds/pyps/conda/py36/bin/conda
CONDA_PREFIX: /reg/neh/home/pennebak/.conda/envs/pre-commit
CONDA_PREFIX_1: /reg/g/pcds/pyps/conda/py36
CONDA_PREFIX_2: /reg/g/pcds/pyps/conda/py36/envs/pcds-3.2.0
CONDA_PROMPT_MODIFIER: (pre-commit) 
CONDA_PYTHON_EXE: /reg/g/pcds/pyps/conda/py36/bin/python
CONDA_ROOT: /reg/g/pcds/pyps/conda/py36
CONDA_SHLVL: 3
PATH: /reg/g/pcds/pyps/conda/py36/bin:/reg/neh/home/pennebak/.conda/envs/pre-commit/bin:/reg/g/pcds/pyps/conda/py36/condabin:/reg/neh/home/pennebak/bin:/reg/g/pcds/epics/extensions/R0.2.0/bin/rhel7-x86_64:/reg/g/pcds/epics/base/R7.0.2-2.0/bin/rhel7-x86_64:/reg/common/tools/bin:/usr/lib64/qt-3.3/bin:/reg/neh/home/pennebak/perl5/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/puppetlabs/bin:/reg/g/pcds/engineering_tools/latest/scripts:/reg/common/tools/bin:/reg/common/tools/bin/eco_tools:/afs/slac/g/lcls/tools/script:/reg/common/tools/bin:/reg/g/pcds/engineering_tools/latest-released/scripts:/reg/neh/operator/xcsopr/bin:/reg/neh/operator/xppopr/bin:/reg/neh/operator/cxiopr/bin:/reg/neh/operator/mfxopr/bin
QT_PLUGIN_PATH: /usr/lib64/kde4/plugins:/usr/lib/kde4/plugins
REQUESTS_CA_BUNDLE: <not set>
SSL_CERT_FILE: <not set>
```
</p>
</details>

**Comment by ZLLentz** [comment]
Seems weird that this problem would be environment-related, since the error is a 404

**Comment by ZryletTC** [comment]
Yeah, I doubt it has anything to do with the environment. Just something about running it on psbuild is different than on my local computer and results in the 404.

**Comment by klauer** [comment]
I was going to try and help debug this, but I can't even create a new pre-commit environment on psbuild-rhel7-01:

```bash
$ conda create -vv -n pre-commit python=3.7 pre-commit

...

DEBUG conda.gateways.subprocess:subprocess_call(46): executing>> "/reg/neh/home/klauer/mc/envs/pre-commit/bin/python3.7" -Wi -m py_compile "/reg/neh/home/klauer/mc/envs/pre-commit/lib/python3.7/site-packages/nodeenv.py"
DEBUG conda.core.link:_execute_actions(562): Error in action #11 for pkg_idx #17 CompilePycAction(target_prefix='/reg/neh/home/klauer/mc/envs/pre-commit', target_short_path='lib/python3.7/site-packages/__pycache__/nodeenv.cpython-37.pyc', source_prefix='/reg/neh/home/klauer/mc/envs/pre-commit', source_short_path='lib/python3.7/site-packages/nodeenv.py', prefix_path_data=PathDataV1(_path='lib/python3.7/site-packages/__pycache__/nodeenv.cpython-37.pyc', path_type='pyc_file'), _execute_successful=False, _verified=True)
Traceback (most recent call last):
  File "/reg/neh/home/klauer/mc/lib/python3.7/site-packages/conda/core/link.py", line 556, in _execute_actions
    action.execute()
  File "/reg/neh/home/klauer/mc/lib/python3.7/site-packages/conda/core/path_actions.py", line 508, in execute
    compile_pyc(python_full_path, self.source_full_path, self.target_full_path)
  File "/reg/neh/home/klauer/mc/lib/python3.7/site-packages/conda/gateways/disk/create.py", line 344, in compile_pyc
    result = subprocess_call(command, raise_on_error=False)
  File "/reg/neh/home/klauer/mc/lib/python3.7/site-packages/conda/gateways/subprocess.py", line 47, in subprocess_call
    p = Popen(command_arg, cwd=path, stdin=PIPE, stdout=PIPE, stderr=PIPE, env=env)
  File "/reg/neh/home/klauer/mc/lib/python3.7/subprocess.py", line 756, in __init__
    restore_signals, start_new_session)
  File "/reg/neh/home/klauer/mc/lib/python3.7/subprocess.py", line 1499, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/reg/neh/home/klauer/mc/envs/pre-commit/bin/python3.7': '/reg/neh/home/klauer/mc/envs/pre-commit/bin/python3.7'
INFO conda.core.link:_reverse_actions(590): ===> REVERSING PACKAGE LINK: conda-forge::nodeenv-1.3.5-py_0 <===
  prefix=/reg/neh/home/klauer/mc/envs/pre-commit
```

OK, forget that. Switching back to a pre-existing environment, I can reproduce:

```bash
$ pre-commit run --all-files
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
return code: 128
expected return code: 0
stdout: (none)
stderr:
    error: RPC failed; result=22, HTTP code = 404
    fatal: The remote end hung up unexpectedly

Check the log at /reg/neh/home/klauer/.cache/pre-commit/pre-commit.log
```

Git clone itself reflects that as well, so it's not pre-commit related:
```bash
$ git clone https://gitlab.com/pycqa/flake8
Cloning into 'flake8'...
error: RPC failed; result=22, HTTP code = 404
fatal: The remote end hung up unexpectedly
```

OK, this led me back to:
https://github.com/pre-commit/pre-commit/issues/1206

Which led me to try:

```bash
$ git clone https://gitlab.com/pycqa/flake8.git
Cloning into 'flake8'...
remote: Enumerating objects: 933, done.
remote: Counting objects: 100% (933/933), done.
remote: Compressing objects: 100% (336/336), done.
remote: Total 9246 (delta 601), reused 835 (delta 510), pack-reused 8313
Receiving objects: 100% (9246/9246), 2.51 MiB | 0 bytes/s, done.
Resolving deltas: 100% (5837/5837), done.
```

😲 

TLDR; add on `.git` to https gitlab URLs
